### PR TITLE
Warnings for unsaved changes, 'intelligent' buttons when editing

### PR DESCRIPTION
**Pull request by shamoon**
This PR addresses #57 and adds warnings when reloading / navigating away from unsaved changes as well as activates / deactivates "Discard" / "Save" buttons only after changes have been made. Notes:

- Uses a very small library for 'dirty' checking: https://github.com/ngneat/dirty-check-forms
- Implemented for editing documents (detail) as well as changes to settings

@jonaswinkler Implementing this does make certain parts of the code a little odd because data is now put in a 'store' and then patched into the forms but Im happy with the result and I think it workasdas well. I've been burned by not having this feature yet, hopefully others find it useful =)

https://user-images.githubusercontent.com/4887959/105815400-b2256f00-5f67-11eb-8f99-7398f62c51fb.mov

### CODE REVIEW

**Comment by jonaswinkler** [comment]
Documents do not loose changes when you navigate away from there, so there's not need to have a dialog at that point.

**Comment by shamoon** [comment]
Hmm, I see your point but that library is designed for checking _dirty_ in comparison to form controls, after looking at this for a while Im not sure how (or even if) doing it in open documents service would work.

_But_, I pushed a commit that does address what you said and only shows the warning when closing a doc, I also added support for warning if there are unsaved changes when someone hits "Close all".

Again, this sounds like a different approach that what you had in mind, take a look when you have a chance and let me know...

File: src-ui/src/app/components/manage/settings/settings.component.ts [hunk 1]
```diff
@@ -45,17 +51,49 @@ export class SettingsComponent implements OnInit {
   ngOnInit() {
     this.savedViewService.listAll().subscribe(r => {
       this.savedViews = r.results
+      let storeData = {
+        'bulkEditConfirmationDialogs': this.settings.get(SETTINGS_KEYS.BULK_EDIT_CONFIRMATION_DIALOGS),
+        'bulkEditApplyOnClose': this.settings.get(SETTINGS_KEYS.BULK_EDIT_APPLY_ON_CLOSE),
```
Comment (jonaswinkler) [inline src-ui/src/app/components/manage/settings/settings.component.ts:56]: Codacy has a fix for the issue: [Strings must use doublequote.](https://app.codacy.com/gh/jonaswinkler/paperless-ng/pullRequest?prid=7592507)
```suggestion
        "bulkEditApplyOnClose": this.settings.get(SETTINGS_KEYS.BULK_EDIT_APPLY_ON_CLOSE),
```

File: src-ui/src/app/components/manage/settings/settings.component.ts [hunk 2]
```diff
@@ -45,17 +51,49 @@ export class SettingsComponent implements OnInit {
   ngOnInit() {
     this.savedViewService.listAll().subscribe(r => {
       this.savedViews = r.results
+      let storeData = {
+        'bulkEditConfirmationDialogs': this.settings.get(SETTINGS_KEYS.BULK_EDIT_CONFIRMATION_DIALOGS),
```
Comment (jonaswinkler) [inline src-ui/src/app/components/manage/settings/settings.component.ts:55]: Codacy has a fix for the issue: [Strings must use doublequote.](https://app.codacy.com/gh/jonaswinkler/paperless-ng/pullRequest?prid=7592507)
```suggestion
        "bulkEditConfirmationDialogs": this.settings.get(SETTINGS_KEYS.BULK_EDIT_CONFIRMATION_DIALOGS),
```

File: src-ui/src/app/components/app-frame/app-frame.component.ts [hunk 3]
```diff
@@ -78,17 +76,21 @@ export class AppFrameComponent implements OnInit, OnDestroy {
...
-    // TODO: is there a better way to do this?
-    let route = this.activatedRoute
-    while (route.firstChild) {
-      route = route.firstChild
-    }
-    if (route.component == DocumentDetailComponent) {
-      this.router.navigate([""])
-    }
+    // user may need to confirm losing unsaved changes
+    this.openDocumentsService.closeAll().subscribe(confirmed => {
```
Comment (jonaswinkler) [inline src-ui/src/app/components/app-frame/app-frame.component.ts:80]: Codacy has a fix for the issue: [Expected parentheses around arrow function argument.](https://app.codacy.com/gh/jonaswinkler/paperless-ng/pullRequest?prid=7592507)
```suggestion
    this.openDocumentsService.closeAll().subscribe((confirmed) => {
```

File: src-ui/src/app/services/open-documents.service.ts [hunk 4]
```diff
@@ -53,6 +57,11 @@ export class OpenDocumentsService {
     }
   }
 
+  setDirty(documentId: number, dirty: boolean) {
+    if (dirty) this.dirtyDocuments.add(documentId)
```
Comment (jonaswinkler) [inline src-ui/src/app/services/open-documents.service.ts:61]: Codacy has a fix for the issue: [Expected { after 'if' condition.](https://app.codacy.com/gh/jonaswinkler/paperless-ng/pullRequest?prid=7592507)
```suggestion
    if (dirty) {this.dirtyDocuments.add(documentId)}
```

File: src-ui/src/app/guards/dirty-form.guard.ts [hunk 5]
```diff
@@ -0,0 +1,29 @@
+import { Injectable } from '@angular/core';
+import { DirtyCheckGuard } from '@ngneat/dirty-check-forms';
```
Comment (jonaswinkler) [inline src-ui/src/app/guards/dirty-form.guard.ts:2]: Codacy has a fix for the issue: [Strings must use doublequote.](https://app.codacy.com/gh/jonaswinkler/paperless-ng/pullRequest?prid=7592507)
```suggestion
import { DirtyCheckGuard } from "@ngneat/dirty-check-forms";
```

File: src-ui/src/app/components/manage/settings/settings.component.ts [hunk 6]
```diff
@@ -45,17 +51,49 @@ export class SettingsComponent implements OnInit {
...
+        'bulkEditApplyOnClose': this.settings.get(SETTINGS_KEYS.BULK_EDIT_APPLY_ON_CLOSE),
+        'documentListItemPerPage': this.settings.get(SETTINGS_KEYS.DOCUMENT_LIST_SIZE),
```
Comment (jonaswinkler) [inline src-ui/src/app/components/manage/settings/settings.component.ts:57]: Codacy has a fix for the issue: [Strings must use doublequote.](https://app.codacy.com/gh/jonaswinkler/paperless-ng/pullRequest?prid=7592507)
```suggestion
        "documentListItemPerPage": this.settings.get(SETTINGS_KEYS.DOCUMENT_LIST_SIZE),
```

File: src-ui/src/app/guards/dirty-form.guard.ts [hunk 7]
```diff
@@ -0,0 +1,29 @@
...
+import { Observable, Subject } from 'rxjs';
```
Comment (jonaswinkler) [inline src-ui/src/app/guards/dirty-form.guard.ts:3]: Codacy has a fix for the issue: [Strings must use doublequote.](https://app.codacy.com/gh/jonaswinkler/paperless-ng/pullRequest?prid=7592507)
```suggestion
import { Observable, Subject } from "rxjs";
```

File: src-ui/src/app/app-routing.module.ts [hunk 8]
```diff
@@ -11,6 +11,7 @@ import { SettingsComponent } from './components/manage/settings/settings.compone
 import { TagListComponent } from './components/manage/tag-list/tag-list.component';
 import { NotFoundComponent } from './components/not-found/not-found.component';
 import { SearchComponent } from './components/search/search.component';
+import { DirtyFormGuard } from './guards/dirty-form.guard';
```
Comment (jonaswinkler) [inline src-ui/src/app/app-routing.module.ts:14]: Codacy has a fix for the issue: [Strings must use doublequote.](https://app.codacy.com/gh/jonaswinkler/paperless-ng/pullRequest?prid=7592507)
```suggestion
import { DirtyFormGuard } from "./guards/dirty-form.guard";
```

File: src-ui/src/app/guards/dirty-form.guard.ts [hunk 9]
```diff
@@ -0,0 +1,29 @@
...
+import { map } from 'rxjs/operators';
+import { NgbModal } from '@ng-bootstrap/ng-bootstrap';
+import { ConfirmDialogComponent } from 'src/app/components/common/confirm-dialog/confirm-dialog.component';
```
Comment (jonaswinkler) [inline src-ui/src/app/guards/dirty-form.guard.ts:6]: Codacy has a fix for the issue: [Strings must use doublequote.](https://app.codacy.com/gh/jonaswinkler/paperless-ng/pullRequest?prid=7592507)
```suggestion
import { ConfirmDialogComponent } from "src/app/components/common/confirm-dialog/confirm-dialog.component";
```

File: src-ui/src/app/services/open-documents.service.ts [hunk 10]
```diff
@@ -2,6 +2,9 @@ import { Injectable } from '@angular/core';
 import { PaperlessDocument } from '../data/paperless-document';
 import { OPEN_DOCUMENT_SERVICE } from '../data/storage-keys';
 import { DocumentService } from './rest/document.service';
+import { NgbModal } from '@ng-bootstrap/ng-bootstrap';
```
Comment (jonaswinkler) [inline src-ui/src/app/services/open-documents.service.ts:5]: Codacy has a fix for the issue: [Strings must use doublequote.](https://app.codacy.com/gh/jonaswinkler/paperless-ng/pullRequest?prid=7592507)
```suggestion
import { NgbModal } from "@ng-bootstrap/ng-bootstrap";
```

File: src-ui/src/app/components/manage/settings/settings.component.ts [hunk 11]
```diff
@@ -1,35 +1,41 @@
...
+    'documentListItemPerPage': new FormControl(null),
+    'darkModeUseSystem': new FormControl(null),
+    'darkModeEnabled': new FormControl(null),
+    'useNativePdfViewer': new FormControl(null),
     'savedViews': this.savedViewGroup,
-    'displayLanguage': new FormControl(this.settings.getLanguage()),
-    'dateLocale': new FormControl(this.settings.get(SETTINGS_KEYS.DATE_LOCALE)),
-    'dateFormat': new FormControl(this.settings.get(SETTINGS_KEYS.DATE_FORMAT)),
+    'displayLanguage': new FormControl(null),
+    'dateLocale': new FormControl(null),
```
Comment (jonaswinkler) [inline src-ui/src/app/components/manage/settings/settings.component.ts:29]: Codacy has a fix for the issue: [Strings must use doublequote.](https://app.codacy.com/gh/jonaswinkler/paperless-ng/pullRequest?prid=7592507)
```suggestion
    "dateLocale": new FormControl(null),
```

**Pull request closed without merging**